File: app/android/app/src/main/java/com/localghost/app/net/EnrollLink.kt
package com.localghost.app.net

import java.net.URLDecoder

/**
 * The self-contained enrollment descriptor a box puts in its QR. No server is involved in
 * discovery or trust: the QR carries everything the phone needs to reach the box on the LAN and
 * to pin its identity on first contact.
 *
 *   localghost://enroll?v=1&host=192.168.1.20&port=8443&fp=AB:CD:...&name=xyntai&cert=<b64url>&key=<b64url>
 *
 * - v     : format version. Absent means 1 (a hand-typed link may omit it). The phone understands up
 *           to CURRENT_VERSION; a higher number means the box is newer than the app, and parse
 *           returns a Result.Outdated so the scanner can say "update the app". cert+key are part of
 *           v1: this is the first published format, nothing older exists in the wild.
 * - host  : LAN address or .local name the phone can route to right now.
 * - port  : the box's mTLS port.
 * - fp    : the box's certificate SHA-256 fingerprint. The trust anchor the phone pins on first
 *           contact, so enrollment is safe even over a hostile network. This fingerprint IS the vouch.
 * - name  : optional human label for the box (prefills the suggested device/box name).
 * - cert  : the DEVICE certificate (client cert) the box issued for this phone, raw DER, base64url-
 *           encoded. DER rather than PEM: PEM is base64 already, so base64url over PEM cost ~1.78x
 *           the DER and pushed a real identity past what the box's QR encoder can hold; DER-direct
 *           is ~1.33x and fits. DeviceCert consumed DER anyway (its PEM path stripped the armour
 *           before parsing), so nothing of value was lost.
 * - key   : that cert's PKCS8 PRIVATE KEY, raw DER, base64url-encoded. The box generates the keypair
 *           (the phone does not) and delivers both here, screen-to-camera , so enrollment is one scan
 *           with NO network call. The box never writes the key to its disk and zeroes its in-memory
 *           copy after rendering the QR; the phone imports these into its encrypted store
 *           (DeviceCert) and presents the cert on every later call. base64url (RFC 4648, URL-safe
 *           alphabet) so the binary survives the query string untouched.
 *
 * Remote access (away from home) is out of scope for the QR: the user types their own DDNS or VPN
 * host into the setup screen. The QR is the LAN/first-contact path.
 *
 * Parsing is pure (no android.net.Uri) so it can be unit-tested. cert/key are OPTIONAL at parse time
 * (a hand-typed link lacks them) but REQUIRED to actually enrol , the enrol flow rejects a link without
 * them, so a code-only link cannot silently half-enrol.
 */
data class EnrollLink(
    val host: String,
    val port: Int,
    val certFingerprint: String,
    val boxName: String = "",
    val version: Int = CURRENT_VERSION,
    // The box-issued device cert + its private key, raw DER, delivered in the QR. Null when the
    // link carried neither (hand-typed path); the enrol flow requires both. Note: ByteArray fields
    // make the data-class equals/hashCode compare these by REFERENCE; nothing compares EnrollLinks
    // for equality today, and content comparison belongs to tests (assertArrayEquals).
    val deviceCertDer: ByteArray? = null,
    val deviceKeyDer: ByteArray? = null,
) {
    /** The base URL the phone will store and connect to. */
    fun baseUrl(): String = "https://$host:$port"

    /** The outcome of parsing a scanned/typed payload. */
    sealed interface Result {
        data class Ok(val link: EnrollLink) : Result
        /** Looked like an enrol link but was malformed (missing required field, bad port). */
        object Malformed : Result
        /** A valid enrol link from a NEWER box than this app understands. Tell the user to update. */
        data class Outdated(val sawVersion: Int) : Result
        /** Not an enrol link at all (wrong scheme). */
        object NotEnroll : Result
    }

    companion object {
        const val PREFIX = "localghost://enroll?"
        /** The newest enrol-link version this app understands. Bump when the format changes. v1 is
         *  the first published format and includes cert+key. */
        const val CURRENT_VERSION = 1

        /**
         * Parse a scanned/typed payload, keeping the version distinction. Use this where the caller
         * wants to tell "newer box, update the app" apart from "garbage".
         */
        fun parseResult(raw: String): Result {
            val text = raw.trim()
            val lower = text.lowercase()
            if (!lower.startsWith(PREFIX)) return Result.NotEnroll
            val params = parseQuery(text.substring(PREFIX.length))

            // version first: absent = 1 (original codes). A higher version than we know means the box
            // is ahead of the app, which is a clean "update" message rather than a parse failure.
            val version = params["v"]?.trim()?.toIntOrNull() ?: 1
            if (version > CURRENT_VERSION) return Result.Outdated(version)

            val host = params["host"]?.trim().orEmpty()
            val fp = params["fp"]?.trim().orEmpty()
            val port = params["port"]?.trim()?.toIntOrNull() ?: 8443
            val name = params["name"]?.trim().orEmpty()
            // The device cert + key, base64url-encoded DER. Optional at parse time (hand-typed links
            // lack them); the enrol flow requires them. Malformed base64 -> treated as absent, not a hard fail.
            val certDer = decodeB64Url(params["cert"]?.trim().orEmpty())
            val keyDer = decodeB64Url(params["key"]?.trim().orEmpty())

            if (host.isEmpty() || fp.isEmpty()) return Result.Malformed
            if (port !in 1..65535) return Result.Malformed

            return Result.Ok(
                EnrollLink(
                    host = host, port = port,
                    certFingerprint = normaliseFp(fp), boxName = name, version = version,
                    deviceCertDer = certDer, deviceKeyDer = keyDer,
                )
            )
        }

        /** Parse a scanned/typed payload. Returns null if it isn't a valid, current enroll link. */
        fun parse(raw: String): EnrollLink? =
            (parseResult(raw) as? Result.Ok)?.link

        private fun parseQuery(query: String): Map<String, String> =
            query.split("&").mapNotNull { pair ->
                val i = pair.indexOf('=')
                if (i <= 0) return@mapNotNull null
                val k = pair.substring(0, i)
                val v = pair.substring(i + 1)
                runCatching {
                    URLDecoder.decode(k, "UTF-8") to URLDecoder.decode(v, "UTF-8")
                }.getOrNull()
            }.toMap()

        /** Normalise a fingerprint to uppercase colon-separated hex for stable comparison. */
        fun normaliseFp(fp: String): String =
            fp.uppercase().filter { it.isLetterOrDigit() }
                .chunked(2).joinToString(":")

        /** Decode a base64url (RFC 4648, URL-safe, optional padding) value to its raw bytes , DER is
         *  binary, so NO UTF-8 step (that would corrupt it) , or null if empty/invalid. Pure JVM
         *  (java.util.Base64) so it stays unit-testable. */
        private fun decodeB64Url(v: String): ByteArray? {
            if (v.isEmpty()) return null
            return runCatching {
                java.util.Base64.getUrlDecoder().decode(v.trimEnd('='))
            }.getOrNull()
        }
    }
}

File: server/internal/pair/enrolllink.go
package pair

import (
	"encoding/base64"
	"fmt"
	"net/url"
	"strings"
)

// EnrollLink is the box-side counterpart to the app's EnrollLink. The string it produces is what
// the QR carries and what EnrollLink.parse on the phone consumes, so the format is a contract:
//
//	localghost://enroll?v=1&host=...&port=...&fp=...&name=...&cert=<b64url>&key=<b64url>
//
// host and fp are mandatory. Without the fingerprint the phone refuses the link, since the
// fingerprint is the trust anchor (no server vouches for the box). There is no pairing code: the
// old code-for-cert exchange is gone (the QR carries the cert), and a code nothing redeems would be
// security theater , a secret in every link that no one checks.
//
// cert and key carry the box-issued DEVICE certificate and its private key: raw DER, base64url-
// encoded (RFC 4648 URL-safe alphabet, no padding, so url.Values never percent-encodes them). DER,
// not PEM: PEM is itself base64, so base64url over PEM costs ~1.78x the DER while DER-direct costs
// ~1.33x , the difference between a link that fits the QR encoder and one that does not. The app
// feeds DER straight to CertificateFactory/PKCS8EncodedKeySpec anyway (its PEM path stripped the
// armour first), so PEM in the link was pure inflation. The box generates the keypair, the phone
// does not; delivery is screen-to-camera, so enrolment is one scan with no network call, the key is
// never written to the box's disk, and the in-memory copy is zeroed after the QR is rendered. They
// are optional in the STRING (a hand-typed link omits them) but the app's enrol flow requires both,
// so a link without them cannot silently half-enrol.
//
// Size, with real numbers: P-256 device cert ~380-500 bytes DER (~510-670 b64url) + PKCS8 key 138
// DER (184) + base link ~130. Total ~820-980. Above the encoder's 666-byte level-M ceiling, inside
// the 858-byte level-L fallback for certs the box actually mints (it controls the subject and
// extensions, so it keeps them minimal). A code-only link stays tiny and stays level M.
//
// v is the format version. It MUST match the app's EnrollLink.CURRENT_VERSION. The app treats an
// absent v as 1 (a hand-typed link may omit it), so emitting it is backward-compatible, and it lets
// a newer box tell an older app to update rather than mis-parsing. cert+key are part of v1: this is
// the first published format, so there is no earlier code-only version in the wild. Bump
// CurrentVersion here in lockstep with the app whenever the link format changes.
type EnrollLink struct {
	Host        string
	Port        int
	Fingerprint string
	BoxName     string
	// DeviceCertDER and DeviceKeyDER are the box-issued client certificate and its PKCS8 private
	// key, raw DER bytes. Nil/empty means the link carries no credential (hand-typed path).
	DeviceCertDER []byte
	DeviceKeyDER  []byte
}

// CurrentVersion is the enrol-link format version this box emits. Keep equal to the app's
// EnrollLink.CURRENT_VERSION.
const CurrentVersion = 1

func (e EnrollLink) String() string {
	q := url.Values{}
	q.Set("v", fmt.Sprintf("%d", CurrentVersion))
	q.Set("host", e.Host)
	q.Set("port", fmt.Sprintf("%d", e.Port))
	// Fingerprint without separators keeps the QR payload short; the app re-inserts colons via
	// its normaliseFp. url.Values will percent-encode anything unusual, so plain hex is best.
	q.Set("fp", stripSeparators(e.Fingerprint))
	if e.BoxName != "" {
		q.Set("name", e.BoxName)
	}
	// base64url without padding: the URL-safe alphabet passes through url.Values.Encode untouched
	// (no percent-encoding, which would inflate the QR), and the app's decoder strips padding anyway.
	if len(e.DeviceCertDER) > 0 {
		q.Set("cert", base64.RawURLEncoding.EncodeToString(e.DeviceCertDER))
	}
	if len(e.DeviceKeyDER) > 0 {
		q.Set("key", base64.RawURLEncoding.EncodeToString(e.DeviceKeyDER))
	}
	// url.Values.Encode sorts keys; the app parser is order-independent, so that is fine.
	return "localghost://enroll?" + q.Encode()
}

// stripSeparators reduces a fingerprint to bare uppercase hex (no colons) for a compact QR.
func stripSeparators(fp string) string {
	var b strings.Builder
	for _, r := range strings.ToUpper(fp) {
		if (r >= '0' && r <= '9') || (r >= 'A' && r <= 'F') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// normaliseFp matches the app: uppercase, colon-separated hex pairs.
func normaliseFp(raw string) string {
	var b strings.Builder
	for _, r := range strings.ToUpper(raw) {
		if (r >= '0' && r <= '9') || (r >= 'A' && r <= 'F') {
			b.WriteRune(r)
		}
	}
	hexStr := b.String()
	var pairs []string
	for i := 0; i+1 < len(hexStr); i += 2 {
		pairs = append(pairs, hexStr[i:i+2])
	}
	return strings.Join(pairs, ":")
}
